### ISSUE TITLE
ranges::size on rvalues

### ISSUE THREAD

**Issue by Arghnews** [ericniebler/range-v3#1433]
Should `ranges::size` work when passed an rvalue of type T that does **not** have a `size` member function, but has at least one of:

- a free function `std::size_t size(const T&)` in the same namespace
- `begin()` and `end()` member functions returning iterators (in the example below said iterators are `const char*` ie. calculating the difference between them is constant, this is a sized range etc.)

I'm using `range-v3` 0.10.0 (latest release) and `-std=c++17`, same on gcc 8 and clang 9, ubuntu x64.

Am I wrong to think this should work? If so, what is the reasoning behind this - my guess (if this is the case and this is deliberate) is this is related to being unable to bind views to rvalue containers to remove potential footguns?

Example:

    #include <cassert>
    #include <cstring>
    #include <range/v3/range/primitives.hpp>
    #include <range/v3/iterator/operations.hpp>
    #include <range/v3/view/all.hpp>
    #include <vector>

    namespace A
    {
      struct MyStringLike
      {
        const char* begin() const { return str; }
        const char* begin() { return str; }
        const char* end() const { return str + std::strlen(str); }
        const char* end() { return str + std::strlen(str); }
        const char* str = "hi there McChum";

        // Uncommenting this function **does** allow the code to compile
        // std::size_t size() const
        // {
          // return std::strlen(str);
        // }
      };

      // Commenting out this function or leaving it in
      // still means this code won't compile
      std::size_t size(const MyStringLike& msl)
      {
        return std::strlen(msl.str);
      }
    }

    int main()
    {
      using namespace A;

      auto m = MyStringLike{};
      assert(ranges::distance(m) == 15);
      assert(ranges::size(m) == 15);

      assert(ranges::size(std::vector<int>{1, 2, 3}) == 3);

      assert(ranges::distance(MyStringLike{}) == 15);

      // This won't compile
      ranges::size(MyStringLike{});
    }

`MyStringLike` does seem to satisfy the conditions listed as required for `ranges::size` to work in [primitives.hpp](https://github.com/ericniebler/range-v3/blob/master/include/range/v3/range/primitives.hpp#L144) a few `static_asserts` do confirm, at least for the begin/end requirement I have checked (haven't checked for the free function size).

Hope I'm not missing something obvious, I appreciate the time taken by any responders, cheers!

**Comment by ericniebler** [comment]
The range library poisons the overload set for `size` in a rather unhelpful way. This is super not-ideal, but you can work around the problem by either:

- adding additional overloads for non-const lvalue reference and rvalue reference, or
- Taking `MyStringLike` by value.

It would be nice to find a better way to constrain the `size` overload set to make your (very reasonable) code valid.

**Comment by CaseyCarter** [comment]
It looks like this case falls through the holes in the poison pills that [P2091R0](https://wg21.link/p2091r0) talks about and fixes.

**Comment by ericniebler** [comment]
Whoa, I missed the fancy footwork for `ranges::size` in P2091 the first time around. I don't quite grok how this solves the problem, though. And how can an implementation detect the difference between xvalues and prvalues when deciding whether to promote the xvalue to an lvalue?

**Comment by CaseyCarter** [comment]
P1870 simplified the opt-in for `borrowed_range` and removed the rvalue begin/end overloads; it also simplified `ranges::begin` and `ranges::end` by having them always operate on lvalues even when the arguments are rvalues. P2091 extended the benefits of that second change to the other access CPOs. `ranges::size`, for example, doesn't care at all if its argument is an lvalue or rvalue and doesn't want to encourage program-defined range types to have different `size` behaviors for lvalues and rvalues, so it always operates on an lvalue bound to its argument. Since we always perform ADL with an lvalue, we can properly poison the lookup with `foo(auto&)` and `foo(const auto&)` overloads. 

The example program fails because the provided non-member size takes `const MyStringLike&` which has a worse conversion sequence for a non-`const` lvalue `MyStringLike` than does the `MyStringLike&` parameter deduced for the "old" forwarding-reference poison pill. P2091 noted cases where the forwarding-reference poison pills incorrectly allow ADL calls they should block, this example shows an ADL call the poison pills should allow but don't.

**Comment by ericniebler** [comment]
> Since we always perform ADL with an lvalue, we can properly poison the lookup with `size(auto&)` and `size(const auto&)` overloads.

But in this case, wouldn't the poison-pill `size(auto&)` still be a better match than the `size(const MyStringLike&)` overload since (presumably) the rvalue gets promoted to a non-const lvalue?

**Comment by CaseyCarter** [comment]
> > Since we always perform ADL with an lvalue, we can properly poison the lookup with `size(auto&)` and `size(const auto&)` overloads.
> 
> But in this case, wouldn't the poison-pill `size(auto&)` still be a better match than the `size(const MyStringLike&)` overload since (presumably) the rvalue gets promoted to a non-const lvalue?

Yes, that's correct, and I'm completely wrong about why this compiles in C++20 but not with range-v3. Overload resolution for the ADL `size` bullet fails due to ambiguity, so `ranges::size` should fall through to the `make-unsigned-like(ranges::end(t) - ranges::begin(t))` case. That is in fact what's happening here; I can remove the body from the non-member `size` overload (`std::size_t size(const MyStringLike& msl);` so it's declared-but-not-defined, and the program still compiles and runs.

> 
> ~EDIT: Also, I still don't understand this:~
> 
> > ~let `t` be an lvalue that denotes the same object as `E` if `E` is a glvalue and otherwise denotes the result of applying the temporary materialization conversion ([conv.rval]) to `E`.~
> 
> ~How is this implementable? It seems to imply that the `ranges::size` CPO has some magic way to distinguish xvalues and prvalues. And it also seems to imply that prvalues are forwarded through as xvalues instead of lvalues. I'm very confused.~
> 
> OK, this is just saying that `ranges::size` takes its argument by `T&&`.

Yes.

> It also still seems to be saying that `t` should be an xvalue in `size(t)` when `E` is a prvalue, which as far as I can tell is not implementable, and not what we want anyway. Still confused.

It's "let `t` be an lvalue that denotes ((the same object as `E` if `E` is a glvalue) or (the result of applying the TMC to `E` otherwise))." This was the result of Richard pointing out that "Let `t` be an lvalue that denotes the same object as `E`" is meaningless when `E` is a prvalue since prvalues do not denote objects. `t` is always an lvalue.

**Comment by ericniebler** [comment]
> It's "let `t` be an lvalue that denotes ((the same object as `E` if `E` is a glvalue) or (the result of applying the TMC to `E` otherwise))." 

That is a valid reading. The other, which is the one which I was reading, is: "let `t` be (an lvalue that denotes the same object as `E` if `E` is a glvalue) or (the result of applying the TMC to `E` otherwise)." Seems ambiguous, except for the fact that my parse leads to utter nonsense, I suppose.

**Comment by CaseyCarter** [comment]
> > It's "let `t` be an lvalue that denotes ((the same object as `E` if `E` is a glvalue) or (the result of applying the TMC to `E` otherwise))."
> 
> That is a valid reading. The other, which is the one which I was reading, is: "let `t` be (an lvalue that denotes the same object as `E` if `E` is a glvalue) or (the result of applying the TMC to `E` otherwise)." Seems ambiguous, except for the fact that my parse leads to utter nonsense, I suppose.

The repeated "denotes" in the actual wording was intended to avoid that reading: "an lvalue that denotes blah if meow or denotes foo otherwise." should suggest that we splice at "denotes". I suppose we could clarify the structure with bullets:

> let `t` be an lvalue that denotes 
> * the same object as `E` if `E` is a glvalue, or 
> * the result of applying the TMC to `E` otherwise.

but the Standard really doesn't need more bullets.

I've posted https://github.com/cplusplus/draft/issues/3802 to ask for clarification of this wording (and suggestions for doing so).

**Comment by ericniebler** [comment]
If you change the poison-pill overloads to:

```c++
void size(const auto&) = delete;
template<class T>
  requires (!requires(const T& t) {size(t);})
void size(auto&) = delete;
```

then `ranges::size` will find and select the `size(const MyStringLike& msl)` overload.

https://godbolt.org/z/aY-RkZ

I'm sick and a little punchy, so it's possible this breaks tons of other stuff.
